File: behavioral/visitor/visitor.ts
class Cat {
    private _name: string;
    private _breed: string;
    private _sound: string;

    constructor(name: string, breed: string, sound: string) {
        this._name = name;
        this._breed = breed;
        this._sound = sound;
    }

    public accept(visitor: Visitor): void {
        visitor.visit(this);
    }

    public get name(): string {
        return this._name;
    }

    public get breed(): string {
        return this._breed;
    }

    public get sound(): string {
        return this._sound;
    }
}

interface Visitor {
    visit(cat: Cat): void;
}

class InfoVisitor implements Visitor {
    public visit(cat: Cat): void {
        console.log(`${cat.breed} cat ${cat.name}.`);
    }
}

class SoundVisitor implements Visitor {
    public visit(cat: Cat): void {
        console.log(`${cat.name}: "${cat.sound}"`);
    }
}

let cats: Cat[] = [
    new Cat('Simba', 'Siamese', 'Meow-meow'),
    new Cat('Oscar', 'Maine Coon', 'Meoow')
];

let infoVisitor: Visitor = new InfoVisitor();
let soundVisitor: Visitor = new SoundVisitor();

for (let cat of cats) {
    cat.accept(infoVisitor);
    cat.accept(soundVisitor);
}

File: behavioral/visitor/visitor.cs
using System;

class VisitorPattern
{
    class Cat
    {
        public string name { get; }
        public string breed { get; }
        public string sound { get; }

        public Cat(string name, string breed, string sound)
        {
            this.name = name;
            this.breed = breed;
            this.sound = sound;
        }

        public void Accept(Visitor visitor)
        {
            visitor.Visit(this);
        }
    }

    interface Visitor
    {
        void Visit(Cat cat);
    }

    class InfoVisitor : Visitor
    {
        public void Visit(Cat cat)
        {
            Console.WriteLine("{0} cat {1}.", cat.breed, cat.name);
        }
    }

    class SoundVisitor : Visitor
    {
        public void Visit(Cat cat)
        {
            Console.WriteLine("{0}: \"{1}\"", cat.name, cat.sound);
        }
    }

    static void Main(string[] args)
    {
        Cat[] cats = {
            new Cat("Simba", "Siamese", "Meow-meow"),
            new Cat("Oscar", "Maine Coon", "Meoow")
        };

        Visitor infoVisitor = new InfoVisitor();
        Visitor soundVisitor = new SoundVisitor();

        foreach (Cat cat in cats)
        {
            cat.Accept(infoVisitor);
            cat.Accept(soundVisitor);
        }
    }
}
